implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricsSender.java

```diff
@@ -23,17 +27,125 @@
+    }
+
+    /**
+     * Immutable representation of a payload of metrics to use with a
+     * {@link OtlpMetricsSender}.
+     *
+     * @since 1.15.0
+     */
+    class Request {
+
+        @Nullable
+        private final String address;
+
+        private final Map<String, String> headers;
+
+        private final byte[] metricsData;
+
+        /**
+         * Represents a payload of metrics to be sent.
+         * @param address where to send the metrics
+         * @param metricsData encoded batch of metrics
+         * @param headers metadata to send as headers with the metrics data
+         */
+        public Request(@Nullable String address, Map<String, String> headers, byte[] metricsData) {
+            this.address = address;
+            this.headers = headers;
+            this.metricsData = metricsData;
+        }
+
+        @Nullable
+        public String getAddress() {
+            return address;
+        }
+
+        public byte[] getMetricsData() {
+            return metricsData;
+        }
+
+        public Map<String, String> getHeaders() {
+            return headers;
+        }
+
+        public static Builder builder(byte[] metricsData) {
+            return new Builder(metricsData);
+        }
+
+        @Override
+        public String toString() {
+            return "OtlpMetricsSender.Request for address: " + address + ", headers: " + headers + ", metricsData: "
+                    + new String(metricsData, StandardCharsets.UTF_8);
+        }
+
+        public static class Builder {
+
+            private final byte[] metricsData;
+
+            @Nullable
+            private OtlpMetricsSender sender;
+
+            @Nullable
+            private String address;
+
+            private Map<String, String> headers = Collections.emptyMap();
+
+            private Builder(byte[] metricsData) {
+                this.metricsData = metricsData;
+            }
+
+            private Builder(byte[] metricsData, OtlpMetricsSender sender) {
+                this.metricsData = metricsData;
+                this.sender = sender;
+            }
+
+            public Builder address(String address) {
+                this.address = address;
+                return this;
+            }
+
+            public Builder headers(Map<String, String> headers) {
+                this.headers = headers;
+                return this;
+            }
+
+            public Request build() {
+                return new Request(address, headers, metricsData);
+            }
+
+            /**
+             * Build this request and send it to the associated sender. Throws an
+             * exception if there is no sender associated with this builder.
+             * @throws Exception when sending the request fails, an exception is thrown
+             */
+            public void send() throws Exception {
+                if (sender == null) {
+                    throw new IllegalStateException("A sender has not been set");
```

Comment:
I'm not sure I like this arrangement - there's basically two paths to using the builder. One where you start with passing a sender and end with calling `send` (although you could still call `build` without issue and manually pass the Request to a sender). The other where you don't pass a sender in the beginning and you should only call `build` at the end and not `send`. I was trying to mirror some of the convenience we have with the HttpSender and make the API more fluent.

I'm not sure how worthwhile it is to do all this engineering on this part of the API, which even users making their own `OtlpMetricsSender` won't interact with, since this API should be used within the `OtlpMeterRegistry`'s `publish` method, and we generally expect users should not need to implement that themselves.

Reply:
I'm questioning the same thing. Wouldn't it be easier to allow for the `Request` to be created and then the `OtlpMetricsSender` interface will just be used to pass the `Request` to. It seems a bit too complex at the moment.

But I do like the fluidness there... If we want to make sure the sender is always present, `Request.builder()` should be made private, so we have a guarantee the interface is always present

Reply:
I've updated the PR so the only way to construct a `Request` object is via the sender's `newRequest` method. I left the option to `build` the `Request`, though that too seems superfluous since the request builder is made from a sender instance.